### PR TITLE
Kube config loader

### PR DESCRIPTION
**Pull request by akshaymankar**
I needed to move a few methods to TLSUtils due to circular dependencies,
this will of course break compatibility with previous versions. I will
re-export all the previous functions again.

Apart from that, there is a fair bit of code in here without any tests,
so next thing I will do is start adding unit tests to make sure the
basics are fine.

The OIDC and GCP Auth code is difficult to test at unit level, so we
may have to figure out some way to integration test those.

I noticed that the code has mixed indentation (2 spaces and 4 spaces),
all the new code in this commit is 2 space indented. I did not reindent
4 spaces lines for easier code review. We should do that after this work
is over.

[#2]

### CODE REVIEW

File: kubernetes-client/package.yaml [hunk 1]
```diff
@@ -27,26 +29,41 @@ tests:
     dependencies:
       - kubernetes-client
 extra-source-files:
+  - README.md
```
Comment (akshaymankar) [inline kubernetes-client/package.yaml]:
```suggestion
```

File: kubernetes-client/src/Kubernetes/Client/Auth/GCP.hs [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+fetchToken GCPAuth{..} = do
+  (stdOut, _) <- readProcess_ gcpCmd
+  let credsJSON = Aeson.eitherDecode stdOut
+                    & first Text.pack
+      token =  runJSONPath gcpTokenKey =<< credsJSON
+      expText = runJSONPath gcpExpiryKey =<< credsJSON
+      expiry :: Either Text (Maybe UTCTime)
+      expiry = Just <$> (parseExpiryTime =<< expText)
+  atomically $ writeTVar gcpAccessToken (rightToMaybe token)
+  atomically $ writeTVar gcpTokenExpiry (either (const Nothing) id expiry)
```
Comment (symbiont-joseph-kachmar) [inline kubernetes-client/src/Kubernetes/Client/Auth/GCP.hs]: It's nicer to wrap both of these inside a single transaction block, if possible:
```haskell
atomically $ do
  writeTVar gcpAccessToken (rightToMaybe token)
  writeTVar gcpTokenExpiry (either (const Nothing) id expiry)
```

File: kubernetes-client/src/Kubernetes/Client/Auth/GCP.hs [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+
+getToken :: GCPAuth -> IO (Either Text Text)
+getToken g@(GCPAuth{..}) = getCurrentToken g
+                           >>= maybe (fetchToken g) (return . Right)
+
+getCurrentToken :: GCPAuth -> IO (Maybe Text)
+getCurrentToken (GCPAuth{..}) = do
+  now <- getCurrentTime
+  maybeExpiry <- atomically $ readTVar gcpTokenExpiry
+  maybeToken <- atomically $ readTVar gcpAccessToken
```
Comment (symbiont-joseph-kachmar) [inline kubernetes-client/src/Kubernetes/Client/Auth/GCP.hs]: Consider `readTVarIO` here, as it is much faster than `atomically . readTVar` since it doesn't need to perform a transaction (cf. https://www.stackage.org/haddock/lts-13.28/base-4.12.0.0/GHC-Conc.html#v:readTVarIO )

**Comment by symbiont-joseph-kachmar** [comment]
First off, thank you for the PR! 

I'm maintaining some of my own hacky auth stuff at the moment, but none of it is in any state to be PR'd to this repo, so your work here is much appreciated!

I've added a few inline comments for some simple changes I think might be useful, as well as some additional notes below that are a little more broadly architectural.

---

Regarding some of the error handling, I think the following changes would make these modules nicer:
- remove all calls to `error` and replace them with either explicit `IO` exceptions or `Either`s
- convert the semantically meaningful errors away from simple strings towards proper error ADTs
  - these ADTs can also be given `Exception` instances so that they can be thrown if necessary
  - ADTs are especially important in the context of `Exception`s, as they give the user a means to catch and handle them (versus dealing with strings)

We should also review some of the additional dependencies here to see what we need and what could be elided to keep the transitive dependency footprint down for users of the library.

**Comment by akshaymankar** [comment]
Hi @symbiont-joseph-kachmar, Thanks a lot for the suggestions. I have resolved the TVar reads. Also applied the same in OIDC Auth. 

The `error` suggestion also make sense, I will work on this next to get rid of all them.

For the dependencies, I will try to analyze them and see if any could be removed. 
It bothers me a little that OIDC and GCP auth bring in dependencies which might not be required by everyone, but I am not sure of a way to keep them out and still have a function which can understand all kubeconfigs.

**Comment by akshaymankar** [comment]
Alright, I have removed all the `error` calls and added a few unit tests.

I am wondering about how we can have integration tests for the the Auth stuff as we'd need some infrastructure for that. I don't know who can sponsor that. If nobody can, I think we can do something with the [kubernetes in docker](https://github.com/kubernetes-sigs/kind) on one of the free public CIs, not sure about the feasibility of it. Please let me know if you have any ideas.

**Comment by symbiont-joseph-kachmar** [comment]
`kind` or `k3s` might be do-able for something like this. I haven't thoroughly explored either of these on my own yet, though.

@guoshimin I think this was brought up in another issue as well, do you have familiarity with any of these, yourself?

**Comment by guoshimin** [comment]
I've used minikube before, but not in a CI setting. We can probably take a look at how other language clients do it.

**Comment by erthalion** [comment]
>  I think we can do something with the kubernetes in docker

Just to share, for one of our projects (zalando/postgres-operator) we also use [kind](https://github.com/kubernetes-sigs/kind) for integration tests. I remember there were some limitations, but overall seems like a feasible approach.

**Comment by akshaymankar** [comment]
Ok, I will try to set up using `kind` on either Travis or Circle CI. 
I think we should merge this PR and setup CI in a separate PR. What do y'all think?

**Comment by akshaymankar** [comment]
I went through the dependencies and found that `jwt` and `jose-jwt` (dependency of `oidc-client`) do similar things. So, I removed dependency on `jwt`. Here is the list of direct dependencies that were added in this PR and I think are difficult to remove, but suggestions are welcome:
* [yaml](https://hackage.haskell.org/package/yaml): Required to read kube config files
* [oidc-client](https://hackage.haskell.org/package/oidc-client): Required for OIDC auth
* [hoauth2](https://hackage.haskell.org/package/hoauth2): Required for refreshing tokens using OAuth when using OIDC auth
* [uri-bytestring](https://hackage.haskell.org/package/uri-bytestring): Required to parse URI in OIDC auth, hoauth2  also depends on it.
* [jose-jwt](https://hackage.haskell.org/package/jose-jwt): Required to parse JWT tokens in OIDC Auth to figure out if the token is expired. It is also a dependency of `oidc-client`
* [timerep](https://hackage.haskell.org/package/): Required to parse time in RFC3339 specification for GCP Auth.
* [jsonpath](https://hackage.haskell.org/package/jsonpath): Required to parse `token-key` and `expiry-key` from GCP Auth config. Aeson's internal [JSONPath](https://hackage.haskell.org/package/aeson-1.4.4.0/docs/Data-Aeson-Internal.html#t:JSONPath) is not adequate for this.
* [typed-process](https://hackage.haskell.org/package/typed-process): Required to run command provided by GCP Auth. It is also a dependency of `hoauth2`. I could've used `System.Process`, but as this library provides a better interface and was anyways required, I chose to keep it as a direct dependency.

I haven't added bounds to these libraries in the `package.yaml` yet. I will do that next and make this PR ready for review.

**Comment by akshaymankar** [comment]
Hello all, this is now ready to be reviewed.

**Comment by akshaymankar** [comment]
I thought I'd send another PR for CI with both integration and unit tests. But I realized I had broken compilation on older versions of GHC, so I pushed the changes here. As a result all the unit tests are running on Travis. The latest build is here: https://travis-ci.org/akshaymankar/kubernetes-client-haskell/builds/572600159

I removed the `travis_wait` from `.travis.yml` because the build was failing and it printed nothing on failure. I think it is not required anymore because stack prints way more logs while compiling code. But please let me know if I'm mistaken.

I had to add `stack-8.4.4.yaml` and `stack-8.2.2.yaml` as the LTS snapshots for those GHC versions didn't have all the packages.

File: kubernetes-client/src/Kubernetes/Client/Auth/OIDC.hs [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+  return $ updateClientParams t <$> BS.readFile caFile
+
+parseCAData :: TLS.ClientParams -> Map Text Text -> Maybe (IO (Either String TLS.ClientParams))
+parseCAData t m = do
+  caText <- Map.lookup "idp-certificate-authority-data" m
+  pure . pure
+    $ (B64.decode $ Text.encodeUtf8 caText)
+    >>= updateClientParams t
+
+lookupEither :: (Show key, Ord key) => Map key val -> key -> Either String val
```
Comment (symbiont-joseph-kachmar) [inline kubernetes-client/src/Kubernetes/Client/Auth/OIDC.hs]: All of the `Either String` stuff here makes me a bit uncomfortable; it's the type of thing that would make this code difficult to work with when you wanted to handle and recover from specific errors.

A concrete error type (if there is only one error) or an error ADT (if there's more than one error that could occur together here) would be nicer if there's something that fits.

File: kubernetes-client/src/Kubernetes/Client/Auth/OIDC.hs [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+  refreshTokenTVar <- atomically $ newTVar $ Map.lookup "refresh-token" m
+  return $ do
+    tlsParams <- eitherTLSParams
+    issuerURL <- lookupEither m "idp-issuer-url"
+    clientID <- lookupEither m "client-id"
+    clientSecret <- lookupEither m "client-secret"
+    return OIDCAuth{..}
+
+parseCA :: Map Text Text -> IO (Either String TLS.ClientParams)
+parseCA m = do
```
Comment (symbiont-joseph-kachmar) [inline kubernetes-client/src/Kubernetes/Client/Auth/OIDC.hs]: There are a lot of single letter variable names (e.g. `m :: Map Text Text`) that make it hard to piece together what some of these things are.

It might be helpful for future maintainers if either `Map Text Text` were under a `newtype` that made it descriptive, or if `m` was named such that it's more descriptive.

File: kubernetes-client/src/Kubernetes/Client/Config.hs [hunk 6]
```diff
@@ -1,136 +1,158 @@
...
+data KubeConfigSource = KubeConfigFile FilePath
+                      | KubeConfigCluster
+
+{-|
+  Creates 'NH.Manager' and 'K.KubernetesClientConfig' for a given
+  'KubeConfigSource'. It is recommended that multiple 'kubeClient' invocations
+  across an application share an 'OIDCCache', this makes sure updation of OAuth
+  token is synchronized across all the different clients being used.
+-}
+kubeClient
```
Comment (symbiont-joseph-kachmar) [inline kubernetes-client/src/Kubernetes/Client/Config.hs]: `mkKubeClient` or `mkKubeClientConfig` might be a better name for this, since it's actually creating the client.

File: kubernetes-client/src/Kubernetes/Client/Config.hs [hunk 7]
```diff
@@ -1,136 +1,158 @@
...
+  (tlsParams, cfg) <-
+    case getAuthInfo kubeConfigFile of
+      Left _          -> return (t,c)
+      Right (_, auth) -> applyAuthSettings oidcCache auth (t, c)
+  mgr <- newManager tlsParams
+  return (mgr, cfg)
+kubeClient _ (KubeConfigCluster) = Kubernetes.Client.Config.cluster
+
+-- |Creates 'NH.Manager' and 'K.KubernetesClientConfig' assuming it is being executed in a pod
+cluster :: (MonadIO m, MonadThrow m) => m (NH.Manager, K.KubernetesClientConfig)
```
Comment (symbiont-joseph-kachmar) [inline kubernetes-client/src/Kubernetes/Client/Config.hs]: Same as before, `mkCluster`, `mkClusterClientConfig`, or some variant is probably more descriptive.

File: kubernetes-client/src/Kubernetes/Client/Config.hs [hunk 8]
```diff
@@ -1,136 +1,158 @@
...
+  -> IO (NH.Manager, K.KubernetesClientConfig)
+kubeClient oidcCache (KubeConfigFile f) = do
+  kubeConfigFile <- decodeFileThrow f
+  uri <- getCluster kubeConfigFile
+         & fmap server
+         & either (const $ pure "localhost:8080") return
+  t <- defaultTLSClientParams
+       & fmap (tlsValidation kubeConfigFile)
+       & (>>= (addCACertData kubeConfigFile))
+       & (>>= addCACertFile kubeConfigFile (takeDirectory f))
```
Comment (symbiont-joseph-kachmar) [inline kubernetes-client/src/Kubernetes/Client/Config.hs]: Some of this partial application makes it hard to follow what's going on here, would it be possible to break some of this stuff out into the main `do` block and to reduce some of the combinator usage?

**Comment by symbiont-joseph-kachmar** [comment]
Okay, I've added some comments but I'd like to qualify them here a bit:

Most of the things I've raised are _very_ superficial (e.g. function/variable naming, partial application, combinator usage) and I think it would be unfair to block this PR over any of them, especially given that this is a large amount of work and you've been building it up for quite some time.

The only thing that I think it might make sense to address directly here would be some of the `Either String` stuff, but even that could be addressed in a future PR.

To that end, I'm going to give it another read through later to try and get a better handle on some of the implemented functionality, but if this is passing tests and you've used it in another project successfully I think I'm onboard to tentatively approve this.

Any cleanup that we think might be necessary can be done in the future, prior to a new release.

@guoshimin @denibertovic do you have any thoughts here?

(cc-ing @denibertovic because I saw them mention that they were interested in doing some Kubernetes Haskell work on their Twitter feed)

File: kubernetes-client/src/Kubernetes/Client/Auth/OIDC.hs (hunk 4 above)
Comment (akshaymankar) [reply kubernetes-client/src/Kubernetes/Client/Auth/OIDC.hs]: This was very valuable, thanks for pointing it out. I have replcaed most of `Either String` with `Either <some exception>` and `throwM`ed the exception at the last possible moment. Let me know if looks good.

File: kubernetes-client/src/Kubernetes/Client/Auth/OIDC.hs (hunk 5 above)
Comment (akshaymankar) [reply kubernetes-client/src/Kubernetes/Client/Auth/OIDC.hs]: I decided to give the variables more understandable name and in general replaced most of the single letter variables with better names. Hopefully it is easier to follow.

File: kubernetes-client/src/Kubernetes/Client/Config.hs (hunk 6 above)
Comment (akshaymankar) [reply kubernetes-client/src/Kubernetes/Client/Config.hs]: Called it `mkKubeClientConfig`

File: kubernetes-client/src/Kubernetes/Client/Config.hs (hunk 7 above)
Comment (akshaymankar) [reply kubernetes-client/src/Kubernetes/Client/Config.hs]: Called it `mkInClusterClientConfig` as most other language clients also call it this.

File: kubernetes-client/src/Kubernetes/Client/Config.hs (hunk 8 above)
Comment (akshaymankar) [reply kubernetes-client/src/Kubernetes/Client/Config.hs]: I have extracted this out into a function and given each step a name so it is more followable.
I found was doing similar things in one more place, I have fixed that too.

**Comment by akshaymankar** [comment]
Thanks a lot for the suggestions @symbiont-joseph-kachmar!
All of them made sense to me and so I have tried to tackle each one of them in subsequent commits. In process I found that there were a few more places which could be made better, I made them better too. 
Please let me know if you have any more feedback about it.

**Comment by akshaymankar** [comment]
Hello @guoshimin, did you get a chance to look at this?

**Comment by guoshimin** [comment]
Sorry @akshaymankar for dropping the ball on this. I've just not had any time or energy to work on this. Thanks @symbiont-joseph-kachmar for stepping up and giving it the review it deserves. I agree that we should merge this.
